# 01. Importing data libraries

In [1]:
# Import pandas, numpy, and os libaries
import pandas as pd
import numpy as np
import os

# 02. Importing data sets

In [2]:
# Import products.csv and orders_wrangled.csv
path = r'/Users/brucehickling/Library/CloudStorage/Dropbox/Courses/CareerFoundry/Data Immersion/4_Python/02-2023_Instacart Basket Analysis'

In [3]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [4]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [5]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [6]:
df_ords.rename(columns = {'time_of_day_order_made' : 'order_hour_of_day'}, inplace = True)

In [7]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

# 03. Data consitency checks on products.csv

In [8]:
# Create a dataframe
df_test = pd.DataFrame()

In [9]:
# Dealing with mixed data types
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [10]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [11]:
# Check for mixed type data
for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0:
        print (col)

mix


In [12]:
df_test['mix'] = df_test['mix'].astype('str')

In [13]:
# Dealing with missing data
# Finding missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [14]:
# View null values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [15]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [16]:
# Compare rows with/ without null values
df_prods.shape

(49693, 5)

In [17]:
# Create dataframe excluding null values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [18]:
df_prods_clean.shape

(49677, 5)

In [19]:
# Dealing with duplicates
# Look for duplicate values
df_dups = df_prods_clean[df_prods_clean.duplicated()]


In [20]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [21]:
# Compare rows with/ without duplicates
df_prods_clean.shape

(49677, 5)

In [22]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [23]:
df_prods_clean_no_dups.shape

(49672, 5)

In [24]:
# Export checked products.csv file
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'), index = False)

# 04. Data consistency checks on orders_wrangled.csv

In [25]:
# Step 2 - run df.describe() on df_ords and note anything that looks off or should be investigated further
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

__Answer for Step 2:__

__a) The count of the variable 'days_since_prior_order' does not match the count of the other variables__

__b) The variable 'order_number'has a maximum value of 100 for 3.4 million orders__

In [26]:
# Step 3 - look for mixed-type data
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print (col)

__Answer for Step 3:__

__There are no mixed-type data in df_ords

In [27]:
# Step 4 - if you find mixed data, fix it so there is only a single data type per column

__Asnwer for Step 4:__

__There is no mixed-type data to fix.__

In [28]:
# Step 5 - run a check for missing values in your dataframe, report any findings, and propose an explanation for anything you find
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

__Answer for Step 5:__

__The variable 'days_since_prior_order' has 206209 listed null/ missing values. I thought at first that one reason could be that this doesn't represent missing information: that a value '0' is returned for this variable if they user_id just made a purchase that day (and so it would be 0 days since their last order). However, I read further about isnull() to confirm that this function returns a dataframe object where all the values are replaced with a Boolean value, so this couldn't be the case. When you check with the results of df_ords.describe() you can see the the count for the variable 'days_since_prior_order' is exactly 206,209 less than the counts for all other variables. This confirms that this variable does in fact have 206209 missing values.__

In [29]:
# Step 6 - Address the missing values using an appropriate method
# Confirm if there are outliers in the variable

__Answer for Step 6:__

__Given 206209 values represents 6% of the dataframe, I don't think the data should be excluded from the analysis. I believe it is possible to impute the missing values.  The mean of the variable from using df_ords.describe() is 11.11 days with a SD of 9.2 days. 2 SDs is 29.51 days (or no less than 0 days) when the max is 30 days, therefore, >95% is within 2SD and so the presence of outliers is eliminated. Therefore, the mean can be used to impute the missing data.__

In [30]:
# Impute the missing data using the mean
df_ords['days_since_prior_order'].fillna(11.111483, inplace=True)

In [31]:
# Create a new dataframe for the cleaned missing data
df_ords_clean = df_ords[df_ords['days_since_prior_order'].isnull() == False]

In [32]:
# Confirm count of 'days_since_prior_order' matches the count of the other variables
df_ords_clean.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.421083e+06  
mean        1.345202e+01            1.111463e+01  
std         4.226088e+00            8.924952e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            5.000000e+00  
50%         1.300000e+01            8.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [33]:
# Step 7 - run a check for duplicate values in your dataframe
df_ords_dups = df_ords_clean[df_ords_clean.duplicated()]

In [34]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

__Answer to Step 7 and Step 8:__

__There are no duplicates in df_ords_clean

In [37]:
# Check the cleaned dataframe
df_ords_clean.shape

(3421083, 6)

In [36]:
# Step 9 - export the cleaned df_ords_clean as orders_checked.csv
df_ords_clean.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'), index = False)